airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/cdc/PostgresCdcCtidInitializer.java

```diff
@@ -211,10 +201,125 @@ public static List<AutoCloseableIterator<AirbyteMessage>> cdcCtidIteratorsCombin
         LOGGER.warn("Streams are under vacuuming, not going to process WAL");
         return initialSyncCtidIterators;
       }
-
     } catch (final SQLException e) {
       throw new RuntimeException(e);
     }
   }
 
+  static class CdcIterator extends AbstractIterator<AirbyteMessage> implements AutoCloseableIterator<AirbyteMessage> {
+
+    private final Iterator<DebeziumComponent.Output> outputIterator;
+    private final StateManager stateManager;
+    private final Instant emittedAt;
+
+    private Iterator<AirbyteMessage> messageIterator;
+
+    public CdcIterator(Iterator<DebeziumComponent.Output> outputIterator, StateManager stateManager, Instant emittedAt) {
+      this.outputIterator = outputIterator;
+      this.stateManager = stateManager;
+      this.emittedAt = emittedAt;
+      this.messageIterator = nextIterator(outputIterator.next());
+    }
+
+    @Override
+    public void close() {
+      endOfData();
+    }
+
+    @Nullable
+    @Override
+    protected AirbyteMessage computeNext() {
+      if (!messageIterator.hasNext()) {
+        if (!outputIterator.hasNext()) {
+          return endOfData();
+        }
+        messageIterator = nextIterator(outputIterator.next());
+      }
+      return messageIterator.next();
+    }
+
+    private Iterator<AirbyteMessage> nextIterator(final DebeziumComponent.Output output) {
+      return new Iterator<>() {
+
+        private Iterator<DebeziumComponent.Record> recordIterator = output.data().iterator();
+
+        @Override
+        public boolean hasNext() {
+          return recordIterator != null;
+        }
+
+        @Override
+        public AirbyteMessage next() {
+          if (recordIterator == null) {
+            throw new NoSuchElementException();
+          }
+          if (recordIterator.hasNext()) {
+            return toAirbyteRecordMessage(recordIterator.next(), emittedAt);
+          } else {
+            recordIterator = null;
+            return toAirbyteStateMessage(stateManager, output);
+          }
+        }
+
+      };
+    }
+
+  }
+
+  static public DebeziumComponent.State toComponentState(StateManager stateManager) {
```

Comment:
> Would this function eventually gets moved into StateManager in the future?
This would induce a tight coupling between the StateManager and the DebeziumComponent. Perhaps that's OK though.

> Also it feels like the implementation could be applied to all other connectors and there's no specific postgres related logic - I wonder if this should go into CDK?

Yes, probably in some utils package outside of the component itself.